drv/transceivers-server/src/main.rs

```diff
@@ -112,6 +158,178 @@ impl ServerImpl {
     }
 }
 
+impl ServerImpl {
+    /// Returns the temperature from a CMIS transceiver.
+    ///
+    /// `port` is a logical port index, i.e. 0-31.
+    fn read_cmis_temperature(
+        &self,
+        port: LogicalPort,
+    ) -> Result<Celsius, FpgaError> {
+        const CMIS_TEMPERATURE_MSB: u8 = 14; // CMIS, Table 8-9
+        self.read_temperature_from_i16(port, CMIS_TEMPERATURE_MSB)
+    }
+
+    /// Returns the temperature from a SFF-8636 transceiver.
+    ///
+    /// `port` is a logical port index, i.e. 0-31.
+    fn read_sff8636_temperature(
+        &self,
+        port: LogicalPort,
+    ) -> Result<Celsius, FpgaError> {
+        const SFF8636_TEMPERATURE_MSB: u8 = 22; // SFF-8636, Table 6-7
+        self.read_temperature_from_i16(port, SFF8636_TEMPERATURE_MSB)
+    }
+
+    /// Trigger a read from the given port's given register, which is assumed to
+    /// be an `i16` containing 1/256 °C.
+    fn read_temperature_from_i16(
+        &self,
+        port: LogicalPort,
+        reg: u8,
+    ) -> Result<Celsius, FpgaError> {
+        self.transceivers.setup_i2c_read(reg, 2, port.as_mask())?;
+
+        #[derive(Copy, Clone, FromBytes, AsBytes)]
+        #[repr(C)]
+        struct StatusAndTemperature {
+            status: u8,
+            temperature: zerocopy::I16<zerocopy::BigEndian>,
+        }
+
+        loop {
+            let mut out = StatusAndTemperature::new_zeroed();
+            self.transceivers
+                .get_i2c_status_and_read_buffer(port, out.as_bytes_mut())?;
+            if out.status & Reg::QSFP::PORT0_STATUS::BUSY == 0 {
+                if out.status & Reg::QSFP::PORT0_STATUS::ERROR != 0 {
+                    return Err(FpgaError::ImplError(0));
+                } else {
+                    // "Externally measured free side device temperatures are
+                    // represented as a 16-bit signed twos complement value in
+                    // increments of 1/256 degrees Celsius"
+                    return Ok(Celsius(out.temperature.get() as f32 / 256.0));
+                }
+            }
+        }
+    }
+
+    fn get_transceiver_interface(
+        &mut self,
+        port: LogicalPort,
+    ) -> Result<ManagementInterface, FpgaError> {
+        self.transceivers.setup_i2c_read(0, 1, port.as_mask())?;
+        let mut out = [0u8; 2]; // [status, SFF8024Identifier]
+
+        // Wait for the I2C transaction to complete
+        loop {
```

Comment:
If this gets stuck it'll starve lower priority tasks. Could it have a max count or start sleeping after some number of retries?